### PR TITLE
Remaining cases from specification

### PR DESCRIPTION
**Pull request by Dveim**

### CODE REVIEW

File: scalameta/src/main/scala/scala/meta/Trees.scala [hunk 1]
```diff
@@ -71,6 +71,7 @@ package scala.meta {
   }
 
   object Ctor {
+    @branch trait Call extends Term
     @branch trait Ref extends Term.Ref
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: `extends Term.Ref with Ctor.Call`

File: scalameta/src/main/scala/scala/meta/Trees.scala [hunk 2]
```diff
@@ -460,9 +483,9 @@ package scala.meta.internal.ast {
                          body: Term) extends Ctor with Stat {
       require(body.isCtorBody)
     }
-    @branch trait Ref extends api.Ctor.Ref with impl.Term.Ref
+    @branch trait Ref extends api.Ctor.Ref with impl.Term.Ref with impl.Ctor.Call
     val Name = Ref.Name
-    type Name = Ref.Name
+    type Name = Ref.Name with Ctor.Call
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: This is unnecessary

File: scalameta/src/main/scala/scala/meta/internal/parsers/Helpers.scala [hunk 3]
```diff
@@ -75,11 +76,13 @@ private[meta] object Helpers {
     }
     def isCtorBody: Boolean = {
       def isSuperCall(tree: Tree): Boolean = tree match {
+        case _: Term.This => true
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Helpers.scala]: Why's this necessary?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 4]
```diff
@@ -50,7 +50,7 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
   def parseStat(): Stat = parseRule(parser => parser.statSeq[Stat] {
     case token if token.is[`import`] => importStmt()
     case token if token.is[`package `] => packageOrPackageObjectDef()
-    case token if token.is[TokenClass.DefIntro] => nonLocalDefOrDcl()
+    case token if token.is[TokenClass.DefIntro] || token.is[Ellipsis] => nonLocalDefOrDcl()
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why can't this check be merged into `is[DefIntro]`?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 5]
```diff
@@ -634,8 +644,8 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
     ts.toList
   }
 
-  @inline final def commaSeparated[T <: Tree : AstMetadata](part: => T): List[T] =
-    tokenSeparated[`,`, T](sepFirst = false, part)
+  @inline final def commaSeparated[T <: Tree : AstMetadata](part: => T, useUnquoteForEllipsis: Boolean = false): List[T] =
+    tokenSeparated[`,`, T](sepFirst = false, part, useUnquoteForEllipsis)
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why's this parameter necessary here?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 6]
```diff
@@ -976,15 +988,28 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
-        case Type.Function(params, res) => atPos(tpe, tpe)(Pat.Type.Function(params.map(p => loop(p.require[Type])), loop(res)))
+        case Type.Function(params, res) => atPos(tpe, tpe)(Pat.Type.Function(params.map {
+          case q @ Type.Arg.Quasi(1, Type.Arg.Quasi(0, _)) => q.become[Pat.Type.Quasi]
+          case p => loop(p.require[Type])
+        }, loop(res)))
         case Type.Tuple(elements) => atPos(tpe, tpe)(Pat.Type.Tuple(elements.map(loop)))
         case Type.Compound(tpes, refinement) => atPos(tpe, tpe)(Pat.Type.Compound(tpes.map(loop), refinement))
         case Type.Existential(underlying, quants) => atPos(tpe, tpe)(Pat.Type.Existential(loop(underlying), quants))
         case Type.Annotate(underlying, annots) => atPos(tpe, tpe)(Pat.Type.Annotate(loop(underlying), annots))
+        case Type.Placeholder(Type.Bounds(None, None)) => convert(tpe)
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why's this clause necessary?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 7]
```diff
@@ -976,15 +988,28 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
+      val t: Type = {
+        val t = if (token.is[`(`]) tupleInfixType() else compoundType()
+        def mkOp(t1: Type) = atPos(t, t1)(Type.ApplyInfix(t, typeName(), t1))
+        token match {
+          case _: `forSome` => next(); atPos(t, t)(Type.Existential(t, existentialStats()))
+          case _: Unquote | _: Ident if !isRawBar => infixTypeRest(mkOp(compoundType()), InfixMode.LeftOp)
+          case _ => t
+        }
+      }
+      loop(t)
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Could you recall the motivation for this change?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 8]
```diff
@@ -1966,11 +1999,12 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
       // simple diagnostics for this entry point
       simplePattern(token => syntaxError("illegal start of simple pattern", at = token))
     def simplePattern(onError: Token => Nothing): Pat = autoPos(token match {
-      case _: Ident | _: `this` =>
-        val isBackquoted = token.code.endsWith("`")
+      case _: Ident | _: `this` | _: Unquote =>
+        val isBackquoted = parser.isBackquoted
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why's there a need to precache this?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 9]
```diff
@@ -1966,11 +1999,12 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
         val sid = stableId()
         val isVarPattern = sid match {
-          case Term.Name(value) => !isBackquoted && value.head.isLower && value.head.isLetter
+          case _: Term.Name.Quasi => false
+          case Term.Name(value) => !isBackquoted && value.head.isLetter && value.head.isLower
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: What was the point in swapping the conditions around? :D

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 10]
```diff
@@ -2007,8 +2042,6 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
           case p => p.require[Pat]
         }
         makeTuple[Pat](patterns, () => Lit.Unit(), Pat.Tuple(_))
-      case _: Unquote =>
-        unquote[Pat]
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Okay, so what did this change buy us?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 11]
```diff
@@ -2111,14 +2149,16 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
         if (mod.hasAccessBoundary) mods.filter(_.hasAccessBoundary).foreach(mod => syntaxError("duplicate access qualifier", at = mod))
       }
       validate()
       mods :+ mod
     }
+    def continueLoop = ahead(token.is[`:`] || token.is[`=`] || token.is[EOF])
     def loop(mods: List[Mod]): List[Mod] = token match {
-      case _: Modifier         => loop(appendMod(mods, modifier()))
+      case _: Unquote => if (continueLoop) mods else loop(appendMod(mods, modifier()))
+      case _: Modifier | _: Ellipsis => loop(appendMod(mods, modifier()))
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Can we build turn this into a tokenclass instead?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 12]
```diff
@@ -2111,14 +2149,16 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
       def validate() = {
         if (isLocal && !mod.tokens.head.is[LocalModifier]) syntaxError("illegal modifier for a local definition", at = mod)
-        mods.foreach(m => if (m.productPrefix == mod.productPrefix) syntaxError("repeated modifier", at = mod))
+        if (!mod.isInstanceOf[Mod.Quasi]) mods.foreach(m => if (m.productPrefix == mod.productPrefix) syntaxError("repeated modifier", at = mod))
         if (mod.hasAccessBoundary) mods.filter(_.hasAccessBoundary).foreach(mod => syntaxError("duplicate access qualifier", at = mod))
       }
       validate()
       mods :+ mod
     }
+    def continueLoop = ahead(token.is[`:`] || token.is[`=`] || token.is[EOF])
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Where does this heuristic come from?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 13]
```diff
@@ -2219,36 +2259,44 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
-          val isExplicitlyLocal = mods.accessBoundary.map(_.isInstanceOf[Term.This]).getOrElse(false)
-          if (ownerIsCase) isExplicitlyLocal
-          else isExplicitlyLocal || (!isValParam && !isVarParam)
+    val name = termName() match {
+      case q: Term.Name.Quasi => q.become[Term.Param.Name.Quasi]
+      case x => x
+    }
+    val tpt =
+      if (token.isNot[`:`])
+        None
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: How do you mean, `None`? Wasn't this method always returning `Some(...)`?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 14]
```diff
@@ -2268,7 +2316,7 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
 
   def typeParam(ownerIsType: Boolean, ctxBoundsAllowed: Boolean): Type.Param = autoPos {
     if (token.is[Unquote]) return unquote[Type.Param]
-    var mods: List[Mod] = annots(skipNewLines = true)
+    var mods: List[Mod] = if (token.is[Ellipsis]) List(ellipsis(1, unquote[Mod])) else annots(skipNewLines = true)
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Is `tparam"@foo ${Mod.Covariant()} T" gonna work?

File: scalameta/src/main/scala/scala/meta/Trees.scala (hunk 1 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: fixed

File: scalameta/src/main/scala/scala/meta/Trees.scala (hunk 2 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: fixed

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 15]
```diff
@@ -2420,7 +2472,7 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
     val isMutable = token.is[`var`]
     next()
     val lhs: List[Pat] = commaSeparated(noSeq.pattern2().require[Pat]).map {
-      // TODO: val `x`: Int should be parsed as Decl.Val(..., Pat.Var.Term(Term.Name("x")), ...)
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why remove this comment?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 16]
```diff
@@ -2658,25 +2710,31 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
     }
-    result
-  }
 
-  def constructorCall(tpe: Type, allowArgss: Boolean = true): Term = {
+  def constructorCall(tpe: Type, allowArgss: Boolean = true): Ctor.Call = {
     object Types {
       def unapply(tpes: Seq[Type.Arg]): Option[Seq[Type]] = {
-        if (tpes.forall(_.isInstanceOf[Type])) Some(tpes.map(_.require[Type]))
+        if (tpes.forall(t => t.isInstanceOf[Type] || t.isInstanceOf[Type.Arg.Quasi])) Some(tpes.map {
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: How can `Type.Arg.Quasi` appear here?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 17]
```diff
@@ -2724,7 +2784,9 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
    */
   def templateParents(): List[Term] = {
     val parents = new ListBuffer[Term]
-    def readAppliedParent() = parents += constructorCall(startModType())
+    def readAppliedParent() =
+      if (token.is[Ellipsis]) parents += ellipsis(1, unquote[Term])
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: `unquote[Term]` or `unquote[Ctor.Call]`?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 18]
```diff
@@ -2827,9 +2891,10 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
     val stats = new ListBuffer[T]
     while (!token.is[StatSeqEnd]) {
       def isDefinedInEllipsis = { if (token.is[`(`] || token.is[`{`]) next(); statpf.isDefinedAt(token) }
-      if (token.is[Ellipsis] && ahead(isDefinedInEllipsis)) stats += ellipsis(1, statpf(token))
-      else if (statpf.isDefinedAt(token)) stats += statpf(token)
-      else if (!token.is[StatSep]) syntaxError(errorMsg, at = token)
+      if (statpf.isDefinedAt(token) || (token.is[Ellipsis] && ahead(isDefinedInEllipsis)))
+        stats += statpf(token)
+      else if (!token.is[StatSep])
+        syntaxError(errorMsg, at = token)
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why this change?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 19]
```diff
@@ -2903,10 +2972,12 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
    */
   def templateStats(): List[Stat] = statSeq(templateStat)
   def templateStat: PartialFunction[Token, Stat] = {
+    case token if token.is[Ellipsis] =>
+      ellipsis(1, unquote[Stat])
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: What about handing `_: Unquote` here in the same way like you handled it in `topStat`?

Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Also, why `token if token.is[Ellipsis]` instead of `_: Ellipsis`?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 5 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Usually ellipsis is parsed with `part`. But sometimes (for example, when parsing paramClauses) it should simple be `ellipsis(1, unquote[...])`, because `part` works ambiguously.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Helpers.scala (hunk 3 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Helpers.scala]:
```
val mods = List(mod"private", mod"final")
val paramss = List(param"x: X", param"x: Y")
val expr = q"this(foo, bar)"
q"..$mods def this(..$paramss) = $expr"
....
[info]   org.scalameta.invariants.InvariantFailedException: invariant failed:
[info] when verifying scala.meta.internal.parsers.Helpers.XtensionTermOps(body).isCtorBody
[info] found that scala.meta.internal.parsers.Helpers.XtensionTermOps(body).isCtorBody is false
[info] where body = this(foo, bar)
```

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 5 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Could you give an example?

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala [hunk 20]
```diff
@@ -345,4 +365,102 @@ class ErrorSuite extends FunSuite {
...
+    assert(typecheckError("""
+      import scala.meta._
+      import scala.meta.dialects.Scala211
+      q"private final def this(x: X, y: Y) = foo"
+    """) === """
+      |<macro>:4: this expected but identifier found
+      |      q"private final def this(x: X, y: Y) = foo"
+      |                                             ^
+    """.trim.stripMargin)
+  }
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: Asking here, since github doesn't show the diff for `QuasiquoteSuite.scala`.

Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: Is there any particular order for tests in QuasiquoteSuite.scala? Would it be possible to order everything according to its order of appearance in quasiquote.md?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 5 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: In this method https://github.com/Dveim/scalameta/blob/master/scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala#L2220-2220.

Here (https://github.com/Dveim/scalameta/blob/master/scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala#L2251) it would parse ellipsis as mods, what is wrong in this case.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Helpers.scala (hunk 3 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Helpers.scala]: This shouldn't work. The correct quasiquote for expr is `ctor"SomeClass(foo, bar)`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 5 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why not have `ellipsis(1, unquote[T])` called every time?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 4 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: This is the only case when `Ellipsis` could (possibly) be merged in `TokenClass`. But that would lead to unnecessary checks inside `DefIntro`, because other parsing situations would become broken with naively added `Ellipsis`. So, I decided to put simplest check here, instead of several complicated in `DefIntro`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Helpers.scala (hunk 3 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Helpers.scala]: Removed that `case`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 7 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: This method didn't work correctly with `(` as token, `forSome`, and probably something else.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 9 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: `Term.Name.Quasi` is `Term.Name`, but trying to extract `value` from first one leads to error.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 10 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: It moved to `case _: Ident | _: `this` | _: Unquote =>`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 5 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Indeed, seems that method could be simplified.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 11 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: I think it is possible, but uniform parsing of Unquotes and Ellipses itself is another refactoring issue. And it requires careful usage analysis (there are much more such cases).

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 6 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Removed.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 12 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: To handle cases like `..@$mods $mod1 x: X ...`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 6 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: But no, it handles `pt"_"`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 13 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Yes, also it didn't handles unquotes. Now it does.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 14 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]:
```
scala> import scala.meta._, scala.meta.dialects.Scala211; implicit class tretgfdsvd(a:Any) { def ===(b:Any) = a == b }
import scala.meta._
import scala.meta.dialects.Scala211
defined class tretgfdsvd

scala> tparam"@foo ${Mod.Covariant()} T"
<console>:16: error: value Covariant is not a member of object scala.meta.Mod
       tparam"@foo ${Mod.Covariant()} T"
                         ^
```

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 15 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Because that todo is done.
```
scala> q"val `x`: Int".show[Raw]
res2: String = Decl.Val(Nil, List(Pat.Var.Term(Term.Name("x"))), Type.Name("Int"))
```

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 16 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: It is necessary to parse `val ctor"(x, ..$tpes) => $tpe" = ctor"(x, y) => z"`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 17 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: `Ctor.Call`, of course. Did that change right before implementing `Ctor.Call` and forgot to review.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 18 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: After some previous changes, first and second cases did same job, so I merged them.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 19 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: 1) `Unquote` is handled within `DefIntro`.
2) To preserve code style in this method.

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala (hunk 20 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: I used that order, more or less. First tests are "legacy" ones and other are (almost) in order with specification. Should I reorder QuasiquoteSuite?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 6 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: patternTyp() called on `_` should give an error, because stuff like `parse 'x match { case y: _ => }'` produces an error. It's impossible to have a Pat.Var.Type at the top level of patternTyp (only as a type argument), and in the same vein it's impossible to have Pat.Type.Wildcard at the top level.

Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Also see https://github.com/scalameta/scalameta/issues/216

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 7 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: You mean, it was impossible to parse `x match { case _: (x, y) => ??? }`?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 9 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: I mean, previously we had `value.head.isLower && value.head.isLetter`. Now we have `value.head.isLetter && value.head.isLower`. What's the reason for this change?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 10 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: You mean, support for stuff like `case Foo[Bar]() => `?

Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Haha!

```
21:33 ~$ parse 'x match { case foo[bar]() => }'
[[syntax trees at end of parser]]// Scala source: tmpR7L1zs
package <empty> {
  class wrapper extends scala.AnyRef {
    def <init>() = {
      super.<init>();
      ()
    };
    x match {
      case foo[(bar @ <empty>)]() => ()
    }
  }
}
```

This means that Pat.Extract.targs should be `Seq[Pat.Type]`, not `Seq[Type]`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 13 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: This should be handled elsewhere, because putting this logic in `param` means that previously invalid code such as `def foo(x) = ???` will now be parsed successfully.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 14 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: I meant `scala.meta.internal.ast.Mod.Covariant()`

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 15 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: It would seem so... I wonder what I had in mind back then.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 18 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Hmm I'm having hard time seeing how `ellipsis(1, statpf(token))` and `statpf(token)` could do the same thing :puzzled:

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala (hunk 20 above)
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: Yes, that would be great. Otherwise, it'd be hard to maintain the file in the future. You did an amazing job writing all these tests btw.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 12 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: This is probably the only thing in this pull request that looked seriously spooky to me. I'd prefer you document it by saying something like "the only things that can come after $mod or $mods are either keywords or $pname/$tpname; the former is easy, but in the case of the latter, we need to take care to not hastily parse those names as modifiers".

Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Also, speaking of those tokens, what happens in the case of `tparam"..$mods $foo[..."` or `tparam"..$mods $foo <: ..."`? Shouldn't `[`, `<:`, `>:` and `<%` be also included into this list?

File: scalameta/src/main/scala/scala/meta/Trees.scala (hunk 1 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: So now, since we have `Ctor.Call`, we can do three things:
1) Update the spec for Template to use `..$ctorcalls` instead of `..$exprs`.
2) Update Template.parents to say `Seq[Ctor.Call]`
3) Update Template to require isCtorCall for all Template.parents (oops, this is already done)

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 6 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Ok, removed that case branch.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 7 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: It was impossible to parse `pt"(X, Y) => Z"`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 9 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Probably copy-pasted from somewhere. Changed.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 10 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: But in Trees we have `class Extract(ref: Term.Ref, targs: Seq[impl.Type], args: Seq[Pat.Arg])`.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 12 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Added comment. Also added these tokens in check expression.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 13 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Checking whether `name` is Unquote should be in this method, I think. Added extra check to avoid situations like in your example.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 14 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]:
```
scala> tparam"@foo ${scala.meta.internal.ast.Mod.Covariant()} T"
<console>:16: error: end of file expected but identifier found
       tparam"@foo ${scala.meta.internal.ast.Mod.Covariant()} T"
                                                              ^

scala> tparam"@foo +T"
res2: scala.meta.Type.Param = @foo +T
```

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 10 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Yeah, that's an oversight.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 13 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Okay, I don't mind, as long as things aren't contradicting the SLS.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 14 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: I think that's a problem, because the first line clearly fits the `tparam"..$mods $tpname[..$tparams] >: $tpeopt <: $tpeopt <% ..$tpes : ..$tpes"` specification.

File: scalameta/src/main/scala/scala/meta/Trees.scala (hunk 1 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: Done.

**Comment by Dveim** [comment]
Moved tests.

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala (hunk 20 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: Moved tests.
